noxfile.py

```diff
@@ -17,20 +24,194 @@
+        for gpi_interface in ("vpi", "vhpi")
+        if (toplevel_lang, gpi_interface) in (("verilog", "vpi"), ("vhdl", "vhpi"))
+    ]
+
+    # Special-case simulators.
+    special = [
+        ("cvc", "verilog", "vpi"),
+        ("ghdl", "vhdl", "vpi"),
+        ("icarus", "verilog", "vpi"),
+        ("questa", "verilog", "vpi"),
+        ("questa", "vhdl", "fli"),
+        ("questa", "vhdl", "vhpi"),
+        ("verilator", "verilog", "vpi"),
+        ("vcs", "verilog", "vpi"),
+    ]
+
+    return standard + special
+
+
+def env_vars_for_test(
+    sim: str, toplevel_lang: str, gpi_interface: str
+) -> Dict[str, str]:
+    """Prepare the environment variables controlling the test run."""
+    e = {}
+    if sim is not None:
+        e["SIM"] = sim
+    if toplevel_lang is not None:
+        e["TOPLEVEL_LANG"] = toplevel_lang
+    assert not (toplevel_lang == "verilog" and gpi_interface != "vpi")
+    if toplevel_lang == "vhdl":
+        e["VHDL_GPI_INTERFACE"] = gpi_interface
+
+    return e
+
+
+def stringify_dict(d: Dict[str, str]) -> str:
+    return ", ".join(f"{k}={v}" for k, v in d.items())
+
+
+#
+# Development pipeline
+#
+# - Use nox to build a sdist; no separate build step is required.
+# - Run tests against the installed sdist.
+# - Collect coverage.
+#
+
+
+@nox.session
+def dev_build(session: nox.Session) -> None:
+    session.warning("No building is necessary for development sessions.")
+
+
+@nox.session
+def dev_test(session: nox.Session) -> None:
+    """Run all development tests as configured through environment variables."""
+
+    # Not using session.notify() here to be able to pass parameters to the
+    # session.
+    dev_test_pytest(session)
+    dev_test_sim(session, sim=None, toplevel_lang=None, gpi_interface=None)
+
 
 @nox.session
-def tests(session: nox.Session) -> None:
-    """run cocotb regression suite"""
+@nox.parametrize("sim,toplevel_lang,gpi_interface", simulator_support_matrix())
+def dev_test_sim(
+    session: nox.Session,
+    sim: Optional[str],
+    toplevel_lang: Optional[str],
+    gpi_interface: Optional[str],
+) -> None:
+    """Test a development version of cocotb against a simulator."""
+
     session.env["CFLAGS"] = "-Werror -Wno-deprecated-declarations -g --coverage"
     session.env["COCOTB_LIBRARY_COVERAGE"] = "1"
     session.env["CXXFLAGS"] = "-Werror"
     session.env["LDFLAGS"] = "--coverage"
-    session.install(*test_deps)
+
+    session.run("pip", "install", *test_deps, *coverage_deps)
     session.install("-e", ".")
-    session.run("pytest")
-    session.run("make", "test", external=True)
+
+    env = env_vars_for_test(sim, toplevel_lang, gpi_interface)
+    config_str = stringify_dict(env)
+
+    # Remove a potentially existing coverage file from a previous run for the
+    # same test configuration.
+    coverage_file = Path(f".coverage.test.sim-{sim}-{toplevel_lang}-{gpi_interface}")
+    try:
+        coverage_file.unlink()
+    except FileNotFoundError:
+        pass
```

Comment:
```suggestion
    with suppress(FileNotFoundError):
        coverage_file.unlink()
```
and `from contextlib import suppress`.